crypto/sm2/sm2_crypt.c

```diff
@@ -391,3 +391,350 @@ int sm2_decrypt(const EC_KEY *key,
 
     return rc;
 }
+
+#ifndef OPENSSL_NO_CNSM
+/* GM/T003_2012 Defined Key Derive Function */
+int kdf_gmt003_2012(unsigned char *out, size_t outlen, const unsigned char *Z, size_t Zlen, const unsigned char *SharedInfo, size_t SharedInfolen, const EVP_MD *md)
+{
+    EVP_MD_CTX *mctx = NULL;
+    unsigned int counter;
+    unsigned char ctr[4];
+    size_t mdlen;
+    int retval = 0;
+
+    if (!out || !outlen) return retval;
+    if (md == NULL) md = EVP_sm3();
+    mdlen = EVP_MD_size(md);
+    mctx = EVP_MD_CTX_new();
+    if (mctx == NULL) {
+        SM2err(SM2_F_KDF_GMT003_2012, ERR_R_MALLOC_FAILURE);
+        goto err;
+    }
+
+    for (counter = 1;; counter++)
+    {
+        unsigned char dgst[EVP_MAX_MD_SIZE];
+
+        EVP_DigestInit(mctx, md);
+        ctr[0] = (unsigned char)((counter >> 24) & 0xFF);
+        ctr[1] = (unsigned char)((counter >> 16) & 0xFF);
+        ctr[2] = (unsigned char)((counter >> 8) & 0xFF);
+        ctr[3] = (unsigned char)(counter & 0xFF);
+        if (!EVP_DigestUpdate(mctx, Z, Zlen))
+            goto err;
+        if (!EVP_DigestUpdate(mctx, ctr, sizeof(ctr)))
+            goto err;
+        if (!EVP_DigestUpdate(mctx, SharedInfo, SharedInfolen))
+            goto err;
+        if (!EVP_DigestFinal(mctx, dgst, NULL))
```

Comment:
Consider squashing these into one _if_ statement:
```
if (!EVP_DigestUpdate(...)
    && !EVP_DigestUpdate(...)
    && !EVP_DigestUpdate(...)
    && ! EVP_DigestFinal(...))
    goto err;
```

Reply:
Maybe it should be '||' instead of '&&'. Because i want goto err if there is anyone is true.